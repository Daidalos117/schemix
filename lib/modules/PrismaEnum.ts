import { PrismaEnumOptions } from "typings/prisma-enum";

export class PrismaEnum {
  private enumMap: Map<string, string | undefined> = new Map();

  constructor(public readonly name: string) {}

  public addValue(value: string, options?: PrismaEnumOptions) {
    this.enumMap.set(value, options?.map);
    return this;
  }

  public toString() {
    const { enumMap } = this;
    const entries = [...enumMap.entries()];

    const padding = entries.reduce((accumulator: number, [{ length }]) => {
      return length > accumulator ? length : accumulator;
    }, 0);

    const lines = entries
      .map(([key, value]) => {
        if (!value) return `  ${key}`;
        return "  " + [key.padEnd(padding), ` @map("${value}")`].join(" ");
      })
      .join("\n");

    return [`enum ${this.name} {`, lines, "}"].join("\n");
  }
}
