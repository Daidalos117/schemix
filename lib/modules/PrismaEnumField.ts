import {
  PrismaFieldAttribute,
  PrismaFieldModifier,
  PrismaFieldTypeName,
} from "typings/prisma-field";

export class PrismaEnumField {
  private attributes: Map<string, PrismaFieldAttribute> = new Map();
  private modifier: PrismaFieldModifier = "";
  private rawAttributeString = "";

  constructor(
    private readonly name: string,
    private type: PrismaFieldTypeName
  ) {}

  public setOptional() {
    this.modifier = "?";
    return this;
  }

  public setList() {
    this.modifier = "[]";
    return this;
  }

  public setUnique() {
    this.attributes.set("unique", "@unique");
    return this;
  }

  public setDefault(defaultValue: string) {
    this.attributes.set("default", `@default(${defaultValue})`);
    return this;
  }

  public mapTo(fieldName: string) {
    this.attributes.set("map", `@map("${fieldName}")`);
    return this;
  }

  public setRawAttributes(rawAttributeString: string) {
    this.rawAttributeString = rawAttributeString;
    return this;
  }

  public toTokenArray() {
    const { name, type, modifier, attributes, rawAttributeString } = this;

    return [
      name,
      type + modifier,
      ...[...attributes.values(), rawAttributeString],
    ] as string[];
  }
}
