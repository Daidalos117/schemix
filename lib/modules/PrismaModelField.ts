import { PrismaFieldAttribute, PrismaFieldModifier, PrismaFieldTypeName } from "@/@types/prisma-field";

export class PrismaModelField {
	private attributes: Map<string, PrismaFieldAttribute> = new Map();
	private modifier: PrismaFieldModifier = "";
	
	constructor(
		private readonly name: string,
		private type: PrismaFieldTypeName
	) {}

	setOptional() {
		this.modifier = "?";		
		return this;
	};

	setList() {
		this.modifier = "[]";
		return this;
	};

	setAsId() {
		this.attributes.set("id", "@id");
		return this;
	};

	setUnique() {
		this.attributes.set("unique", "@unique");
		return this;
	};
	
	setToUpdatedAt() {
		this.attributes.set("updatedAt", "@updatedAt");
		return this;
	};

	setDefault(defaultValue: string | number | { cuid: boolean, uuid: boolean, now: boolean }) {
		const setDefaultValue = (value: string) => this.attributes.set("default", `@default(${value})`);
		switch (typeof defaultValue) {
			case "object":
				const [prismaFunc] = Object.entries(defaultValue).find(([_key, value]) => value) || [];
				setDefaultValue(`${prismaFunc}()`)
				break;
			case "string":
				setDefaultValue(`"${defaultValue}"`)
				break;
			default:
				setDefaultValue(defaultValue.toString())
		}
		
		return this;
	};

	mapTo(fieldName: string) {
		this.attributes.set("map", `@map("${fieldName}")`);
		return this;
	};

	toTokenArray() {
		const { name, type, modifier, attributes } = this;
		return [name, type + modifier, ...attributes.values()] as string[];
	};
};