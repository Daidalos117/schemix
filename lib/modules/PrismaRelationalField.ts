import {
  PrismaFieldAttribute,
  PrismaFieldModifier,
  PrismaFieldTypeName,
} from "@/@types/prisma-field";

export class PrismaRelationalField {
  private relationAttributes: Map<string, string> = new Map();
  private attributes: Map<string, PrismaFieldAttribute> = new Map();
  private modifier: PrismaFieldModifier = "";
  private rawAttributeString = "";

  constructor(
    private readonly name: string,
    private type: PrismaFieldTypeName
  ) {}

  private parseRelationAttribute() {
    const { relationAttributes } = this;
    if (!relationAttributes.size) return;
    const relationString = [...relationAttributes.entries()]
      .map((entries) => entries.join(": "))
      .join(", ");
    return `@relation(${relationString})`;
  }

  public setFields(tokens: string[]) {
    this.relationAttributes.set("fields", `[${tokens.join(", ")}]`);
    return this;
  }

  public setReferences(tokens: string[]) {
    this.relationAttributes.set("references", `[${tokens.join(", ")}]`);
    return this;
  }

  public setOptional() {
    this.modifier = "?";
    return this;
  }

  public setList() {
    this.modifier = "[]";
    return this;
  }

  public mapTo(fieldName: string) {
    this.attributes.set("map", `@map("${fieldName}")`);
    return this;
  }

  public setRawAttributes(rawAttributeString: string) {
    this.rawAttributeString = rawAttributeString;
    return this;
  }

  public toTokenArray() {
    const { name, type, modifier, attributes, rawAttributeString } = this;
    const parsedRelationAttribute = this.parseRelationAttribute();

    return [
      name,
      type + modifier,
      ...(parsedRelationAttribute ? [parsedRelationAttribute] : []),
      ...[...attributes.values(), rawAttributeString],
    ] as string[];
  }
}
