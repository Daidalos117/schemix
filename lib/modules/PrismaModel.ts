import { PrismaSchema } from "@/modules/PrismaSchema";
import { PrismaScalarField } from "@/modules/PrismaScalarField";
import { PrismaRelationalField } from "@/modules/PrismaRelationalField";

import { handleRelationalOptions, handleScalarOptions } from "@/util/options";

import { PrismaFieldTypeName } from "@/@types/prisma-field";
import {
	BooleanFieldOptions,
	DateTimeFieldOptions,
	FieldOptions,
	FloatFieldOptions,
	IntFieldOptions,
	RelationalFieldOptions,
	StringFieldOptions
} from "@/@types/prisma-type-options";

export class PrismaModel {
	private fields: Map<string, PrismaScalarField | PrismaRelationalField> = new Map();
	private rawFields: string[] = [];
	
	constructor(private readonly schema: PrismaSchema, public readonly name?: string) {};

	public string(fieldName: string, options?: StringFieldOptions) {
		return this.createField(fieldName, "String", options);
	};

	public int(fieldName: string, options?: IntFieldOptions) {
		return this.createField(fieldName, "Int", options);
	};

	public float(fieldName: string, options?: FloatFieldOptions) {
		return this.createField(fieldName, "Float", options);
	};

	public boolean(fieldName: string, options?: BooleanFieldOptions) {
		return this.createField(fieldName, "Boolean", options);
	};

	public dateTime(fieldName: string, options?: DateTimeFieldOptions) {
		return this.createField(fieldName, "DateTime", options);
	};

	public relation(fieldName: string, model: PrismaModel, options?: RelationalFieldOptions) {
		return this.createRelation(fieldName, model, options);
	};

	public mixin(model: PrismaModel) {
		[...model.fields.entries()].map(([key, value]) => this.fields.set(key, value));
		return this;
	};
	
	public raw(fieldText: string) {
		this.rawFields.push(fieldText);
	};

	public extend(name: string) {
		const clone = this.schema.createModel(name)
		clone.fields = this.fields;
		clone.rawFields = this.rawFields;
		return clone;
	}
	
	public toString() {
		return [
			`model ${this.name} {`,
			this.parseFields(),
			"}"
		].join("\n");
	};

	private createRelation(fieldName: string, model: PrismaModel, options: RelationalFieldOptions = {}) {
		if (!model.name) return this;
		const field = new PrismaRelationalField(fieldName, model.name);
		handleRelationalOptions(field, options);
		this.fields.set(fieldName, field);
		return this;
	}
	
	private createField(fieldName: string, type: PrismaFieldTypeName, options: FieldOptions = {}) {
		const field = new PrismaScalarField(fieldName, type);
		handleScalarOptions(field, options);
		this.fields.set(fieldName, field);
		return this;
	};

	private parseFields() {
		const fields = [...this.fields.values()].map((field) => field.toTokenArray());
		const mostTokens = Math.max(...fields.map(({ length }) => length))
		const paddings = Array(mostTokens).fill(0);

		for (let i = 0; i < mostTokens; i++)
			for (const tokens of fields)
				if (!tokens[i]) continue
				else
					paddings[i] = tokens[i].length > paddings[i] 
						? tokens[i].length
						: paddings[i];

		return [
			...fields.map((tokens) => {
				return "  " + tokens.map((token, index) => 
					token.padEnd(paddings[index])
				).join(" ").trim();
			}),
			...this.rawFields.map((rawField) => "  " + rawField)
		].join("\n");
	};
};