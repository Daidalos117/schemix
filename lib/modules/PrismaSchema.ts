import { PrismaModel } from "@/modules/PrismaModel";

import { exportSchema } from "@/util/export";
import { parseKeyValueBlock } from "@/util/blocks";

import { PrismaDataSourceOptions } from "@/@types/prisma-datasource";
import { PrismaGeneratorOptions } from "@/@types/prisma-generator";

export class PrismaSchema {
	private models: Map<string, PrismaModel> = new Map();
	
	constructor(
		private readonly datasource: PrismaDataSourceOptions,
		private readonly generator: PrismaGeneratorOptions
	) {};

	private parseDataSource() {
		return parseKeyValueBlock("datasource", "database", Object.entries(this.datasource));
	};
	
	private parseGenerator() {
		return parseKeyValueBlock("generator", "client", Object.entries(this.generator));
	};

	public createMixin() {
		const model = new PrismaModel(this);
		return model;
	};

	public createModel(modelName: string) {
		const model = new PrismaModel(this, modelName);
		this.models.set(modelName, model);
		return model;
	};

	public toString() {
		const models = [
			this.parseDataSource(),
			this.parseGenerator(),
			...this.models.values()
		];
		
		return models.map((model) => model.toString()).join("\n\n");
	};

	public export(filepath: string, filename: string) {
		exportSchema(filepath, filename, this);
	}
};