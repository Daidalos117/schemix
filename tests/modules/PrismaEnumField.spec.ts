import { describe, expect, it } from "vitest";
import { PrismaEnumField } from "schemix/lib/modules/PrismaEnumField";

const fieldTypes = [
  "BigInt",
  "Boolean",
  "DateTime",
  "Decimal",
  "Float",
  "Int",
  "Json",
  "String",
];

describe("PrismaEnumField", () => {
  it("Should support creating each type of field", () => {
    const fields = fieldTypes.map((type) => {
      const name = type.toLowerCase();
      return new PrismaEnumField(name, type);
    });

    const fieldAsString = fields.map((field) => field.toTokenArray());
    expect(fieldAsString).toMatchSnapshot();
  });
  it("Should support making fields optional", () => {
    const field = new PrismaEnumField("points", "Int").setOptional();
    const [, token] = field.toTokenArray();
    expect(token).toBe("Int?");
  });
  it("Should support mapping fields", () => {
    const field = new PrismaEnumField("points", "Int").mapTo("points_map");
    expect(field.toTokenArray()).includes('@map("points_map")');
  });
  it("Should support setting default values", () => {
    const field = new PrismaEnumField("points", "Int").setDefault("3");
    expect(field.toTokenArray()).includes("@default(3)");
  });
  it("Should support setting field to list", () => {
    const field = new PrismaEnumField("points", "Int").setList();
    expect(field.toTokenArray()).includes("Int[]");
  });
  it("Should support setting field to be unique", () => {
    const field = new PrismaEnumField("points", "Int").setUnique();
    expect(field.toTokenArray()).includes("@unique");
  });
  it("Should support setting raw attributes.", () => {
    const field = new PrismaEnumField("points", "Int").setRawAttributes(
      "@database.VarChar(255)"
    );
    expect(field.toTokenArray()).includes("@database.VarChar(255)");
  });
  it("Should support multiple modifiers at once", () => {
    const field = new PrismaEnumField("points", "Int")
      .mapTo("points_map")
      .setDefault("3")
      .setRawAttributes("@database.VarChar(255)");

    expect(field.toTokenArray())
      .includes("@default(3)")
      .and.includes('@map("points_map")')
      .and.includes("@database.VarChar(255)");
  });
});
