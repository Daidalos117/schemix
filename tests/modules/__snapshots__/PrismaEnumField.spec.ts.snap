// Vitest Snapshot v1

exports[`PrismaEnumField > Should support creating each type of field 1`] = `
[
  [
    "bigint",
    "BigInt",
    "",
  ],
  [
    "boolean",
    "Boolean",
    "",
  ],
  [
    "datetime",
    "DateTime",
    "",
  ],
  [
    "decimal",
    "Decimal",
    "",
  ],
  [
    "float",
    "Float",
    "",
  ],
  [
    "int",
    "Int",
    "",
  ],
  [
    "json",
    "Json",
    "",
  ],
  [
    "string",
    "String",
    "",
  ],
]
`;
